File: python/subscribe.py
from .publish import list_topics, create_topic
import os
import sys
import logging

from google.cloud import pubsub_v1  # noqa E402

TESTING = sys.argv[1:2] == ['test']


def list_subscriptions_in_topic(project, topic_name):
    """Lists all subscriptions for a given topic."""
    subscriber = pubsub_v1.SubscriberClient()
    topic_path = subscriber.topic_path(project, topic_name)
    return [subscription for subscription in subscriber.list_subscriptions(topic_path)]


def list_subscriptions_in_project(project):
    """Lists all subscriptions in the current project."""
    subscriber = pubsub_v1.SubscriberClient()
    project_path = subscriber.project_path(project)
    return [subscription for subscription in subscriber.list_subscriptions(project_path)]


def create_subscription(project, topic_name, subscription_name):
    """Create a new pull subscription on the given topic."""
    subscriber = pubsub_v1.SubscriberClient()
    topics = list_topics(project)
    if len([topic for topic in topics if topic.name.split('/')[-1] == topic_name]) == 0:
        create_topic(project, topic_name)
    topic_path = subscriber.topic_path(project, topic_name)
    subscription_path = subscriber.subscription_path(
        project, subscription_name)
    subscription = subscriber.create_subscription(
        subscription_path, topic_path)

    logging.info('Subscription created: {}'.format(subscription))


def listen_for_messages(project, subscription_name, callback=None):
    """Receives messages from a pull subscription."""
    subscriber = pubsub_v1.SubscriberClient()
    subscription_path = subscriber.subscription_path(
        project, subscription_name)

    def default_callback(message):
        logging.info('Received message: {}'.format(message))
        message.ack()
    if callback is None:
        callback = default_callback
    logging.info(subscription_path)
    subscription = subscriber.subscribe(subscription_path)
    future = subscription.open(callback)
    return (future, subscription)


def setup_subscription_for_listening(topic_name, subscription_name, callback=None):
    """Setup subscription process"""
    if TESTING:
        return
    project = os.getenv('GCLOUD_PROJECT', 'undefined')
    full_topic_name = topic_name + '-' + os.getenv('ENV', 'development')
    if os.getenv('ENV', 'development') != 'production':
        full_topic_name += ('-' + os.getenv('LOGNAME', 'undefined') + '-' +
                            os.getenv('TRAVIS_BRANCH', 'undefined') + '-' +
                            os.getenv('POD_NAMESPACE', 'undefined'))
    full_subscription_name = full_topic_name + '-' + subscription_name
    subscriptions = list_subscriptions_in_project(project)
    if len([subscription for subscription in subscriptions
            if subscription.name.split('/')[-1] == full_subscription_name]) == 0:
        create_subscription(project, full_topic_name, full_subscription_name)
    return listen_for_messages(project, full_subscription_name, callback)

File: js/subscribe.js
const PubSub = require("@google-cloud/pubsub");
const { MINUTE, SECOND } = require("time-constants");
const { listAllTopics, createTopic } = require("./publish");

const listSubscriptions = () => {
  // Instantiates a client
  const pubsub = PubSub();

  // Lists all subscriptions in the current project
  return pubsub.getSubscriptions()
    .then((results) => {
      const subscriptions = results[0];

      console.log("Subscriptions:");
      subscriptions.forEach(subscription => console.log(subscription.name));

      return subscriptions;
    });
};

const listTopicSubscriptions = async (topicName) => {
  const allTopics = await listAllTopics();
  let topic = allTopics.find(t => t.name.split("/").slice(-1).pop() === topicName);
  if (!topic) {
    topic = await createTopic(topicName);
  }
  const results = await topic.getSubscriptions();
  return results[0];
};

const createSubscription = async (topicName, subscriptionName) => {
  const allTopics = await listAllTopics();
  let topic = allTopics.find(t => t.name.split("/").slice(-1).pop() === topicName);
  if (!topic) {
    topic = await createTopic(topicName);
  }
  const results = await topic.createSubscription(subscriptionName);
  return results[0];
};

const listenForMessages = (subscriptionName, messageHandler, timeOut) => {
  const pubsub = PubSub();
  const subscription = pubsub.subscription(subscriptionName);
  let messageCount = 0;
  const defaultMessageHandler = (message) => {
    console.log(`Received message ${message.id}:`);
    console.log(`\tData: ${message.data}`);
    console.log(`\tAttributes: ${message.attributes}`);
    messageCount += 1;
    console.log(messageCount);
    message.ack();
  };
  const handler = messageHandler || defaultMessageHandler;
  subscription.on("message", handler);
  if (timeOut) {
    setTimeout(() => {
      subscription.removeListener("message", handler);
      console.log("Stopping Subscription");
    }, timeOut * SECOND);
  }
  subscription.on("error", () => {
    setTimeout(() => {
      subscription.removeListener("message", handler);
      subscription.on("message", handler);
    }, MINUTE);
  });
};

const setupSubscriptionForListening = async (props) => {
  const { topicName, subscriptionName, messageHandler, timeOut } = props;
  let fullTopicName = `${topicName}-${process.env.NODE_ENV || "undefined"}`;
  if ((process.env.NODE_ENV !== "production") || (process.env.POD_NAMESPACE != "live")) {
    fullTopicName = `${fullTopicName}-${process.env.LOGNAME || "undefined"}-${process.env.TRAVIS_BRANCH || "undefined"}-${process.env.POD_NAMESPACE || "undefined"}`;
  }
  const fullSubscriptionName = `${fullTopicName}-${subscriptionName}`;
  const topicSubs = await listTopicSubscriptions(fullTopicName);
  const matchingSubs = topicSubs.filter(topic =>
    topic.name.split("/").slice(-1)[0].toLowerCase() === fullSubscriptionName.toLowerCase());
  if (!matchingSubs.length) {
    await createSubscription(fullTopicName, fullSubscriptionName);
    await listTopicSubscriptions(fullTopicName);
    await listenForMessages(fullSubscriptionName, messageHandler, timeOut);
  } else {
    await listenForMessages(fullSubscriptionName, messageHandler, timeOut);
  }
};

module.exports = {
  listSubscriptions: listSubscriptions,
  listTopicSubscriptions: listTopicSubscriptions,
  createSubscription: createSubscription,
  listenForMessages: listenForMessages,
  setupSubscriptionForListening: setupSubscriptionForListening,
};

